File: solutions/php/phone-number/1/PhoneNumber.php
<?php

declare(strict_types=1);

class PhoneNumber
{
    private string $clean = "";
    
    function __construct(string $dirty) {
        $digitCount = 0;
        $clean = "";

        for ($i = 0; $i < strlen($dirty); $i++){
            if (ctype_digit($dirty[$i])){
                $digitCount++;
                $clean = $clean . $dirty[$i];
            } else if (ctype_alpha($dirty[$i])) {
                throw new \InvalidArgumentException('letters not permitted');
            } else if ($dirty[$i] != '(' && $dirty[$i] != ')' && $dirty[$i] != ' ' && $dirty[$i] != '+' && $dirty[$i] != '-' && $dirty[$i] != '.'  ) {
                throw new \InvalidArgumentException('punctuations not permitted');
            }
        }

        if ($digitCount > 11)
            throw new \InvalidArgumentException('more than 11 digits');
        if ($digitCount < 10)
            throw new \InvalidArgumentException('incorrect number of digits');

        if ($digitCount == 11 && substr($clean, 0, 1) != '1')
            throw new \InvalidArgumentException('11 digits must start with 1');

        $areaCode = $digitCount == 11 ? substr($clean, 1, 3) : substr($clean, 0, 3);
        $exchangeCode = $digitCount == 11 ? substr($clean, 4, 3) : substr($clean, 3, 3);

        if (substr($areaCode, 0, 1) == '0')
            throw new \InvalidArgumentException('area code cannot start with zero');
        if (substr($areaCode, 0, 1) == '1')
            throw new \InvalidArgumentException('area code cannot start with one');
        if (substr($exchangeCode, 0, 1) == '0')
            throw new \InvalidArgumentException('exchange code cannot start with zero');
        if (substr($exchangeCode, 0, 1) == '1')
            throw new \InvalidArgumentException('exchange code cannot start with one');
      
        $this->clean = substr($clean, 0, 1) == '1' ? substr($clean, 1) : $clean;
    }

    public function number(): string
    {
        return $this->clean;
    }
}

File: solutions/python/phone-number/1/phone_number.py
class PhoneNumber:

    @staticmethod
    def validate_area_code(candidate: str):
        if candidate[0] == '0':
            raise ValueError("area code cannot start with zero")
        if candidate[0] == '1':
            raise ValueError("area code cannot start with one")

    @staticmethod
    def validate_exchange_code(candidate: str):
        if candidate[0] == '0':
            raise ValueError("exchange code cannot start with zero")
        if candidate[0] == '1':
            raise ValueError("exchange code cannot start with one")
        
    def __init__(self, number):
        if any(c.isalpha() for c in number):
            raise ValueError("letters not permitted")
        if any(not c.isdigit() and c not in {'(',')','-',' ','.','+'} for c in number):
            raise ValueError("punctuations not permitted")

        candidate = "".join(c for c in number if c.isdigit())
        
        if len(candidate) < 10:
            raise ValueError("must not be fewer than 10 digits")
        elif len(candidate) == 11:
            if candidate[0] != "1":
                raise ValueError("11 digits must start with 1")
            candidate = candidate[1:]
        elif len(candidate) > 11:
            raise ValueError("must not be greater than 11 digits")
                    
        self.validate_area_code(candidate[0:3])
        self.validate_exchange_code(candidate[3:6])

        self.number = candidate
        self.area_code = candidate[:3]

    def pretty(self) -> str:
        return f"({self.area_code})-{self.number[3:6]}-{self.number[6:]}"
